truffle/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/InstrumentationHandler.java

```diff
@@ -0,0 +1,716 @@
+        ProbeNode probeNode = parent.getProbeNode();
+        if (TRACE) {
+            SourceSection section = probeNode.getContext().getInstrumentedSourceSection();
+            trace("Invalidate wrapper for %s, section %s tags %s%n", node, section, Arrays.toString(section.getTags()));
+        }
+        if (probeNode != null) {
+            probeNode.invalidate();
+        }
+    }
+
+    static Instrumentable getInstrumentable(Node node) {
+        Instrumentable instrumentable = node.getClass().getAnnotation(Instrumentable.class);
+        if (instrumentable != null && !(node instanceof WrapperNode)) {
+            return instrumentable;
+        }
+        return null;
+    }
+
+    private abstract class AbstractNodeVisitor implements NodeVisitor {
+
+        abstract boolean shouldVisit(RootNode root);
+
+    }
+
+    private abstract class AbstractBindingVisitor extends AbstractNodeVisitor {
+
+        protected final EventBinding<?> binding;
+
+        AbstractBindingVisitor(EventBinding<?> binding) {
+            this.binding = binding;
+        }
+
+        @Override
+        boolean shouldVisit(RootNode root) {
+            return isInstrumentedRoot(root, binding, root.getSourceSection());
+        }
+
+        public final boolean visit(Node node) {
+            SourceSection sourceSection = node.getSourceSection();
+            if (sourceSection != null) {
+                if (isInstrumentedLeaf(node, binding, sourceSection)) {
+                    if (TRACE) {
+                        trace("Filter hit section:%s tags:%s%n", sourceSection, Arrays.toString(sourceSection.getTags()));
+                    }
+                    visitInstrumented(node, sourceSection);
+                }
+            }
+            return true;
+        }
+
+        protected abstract void visitInstrumented(Node node, SourceSection section);
+
+    }
+
+    private abstract class AbstractBindingsVisitor extends AbstractNodeVisitor {
+
+        private final List<EventBinding<?>> bindings;
+
+        AbstractBindingsVisitor(List<EventBinding<?>> bindings) {
+            this.bindings = bindings;
+        }
+
+        @Override
+        boolean shouldVisit(RootNode root) {
+            SourceSection sourceSection = root.getSourceSection();
+            for (int i = 0; i < bindings.size(); i++) {
+                EventBinding<?> binding = bindings.get(i);
+                if (isInstrumentedRoot(root, binding, sourceSection)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public final boolean visit(Node node) {
+            SourceSection sourceSection = node.getSourceSection();
+            if (sourceSection != null) {
+                List<EventBinding<?>> b = bindings;
+                for (int i = 0; i < b.size(); i++) {
+                    EventBinding<?> binding = b.get(i);
+                    if (isInstrumented(node, binding, sourceSection)) {
```

Comment:
Not every node with SourceSection must be instrumentable. So this might not even be worth a warning. I have instead added the option -Dtruffle.instrumentation.trace=true to trace whats happening. It also shows a special message when something is instrumented but not instrumentable.